### ISSUE TITLE
Package install fines with pip but fails with pdm

### ISSUE THREAD

**Issue by m-legrand** [pdm-project/pdm#2098]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

- Create normal python venv on Windows and `pip install connectorx`: it works
- Create a pdm project and run `pdm add connectorx`: it fails

## Actual behavior

```
  ✖ Install connectorx 0.3.1 failed

ERRORS:
add connectorx failed:
Traceback (most recent call last):
  File "C:\Python310\lib\concurrent\futures\thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "C:\Users\MLegrand\.local\pipx\venvs\pdm\lib\site-packages\pdm\installers\synchronizers.py", line 277, in install_candidate
    self.manager.install(can)
  File "C:\Users\MLegrand\.local\pipx\venvs\pdm\lib\site-packages\pdm\installers\manager.py", line 33, in install
    installer(str(prepared.build()), self.environment, prepared.direct_url())
  File "C:\Users\MLegrand\.local\pipx\venvs\pdm\lib\site-packages\pdm\models\candidates.py", line 375, in build
    self.obtain(allow_all=False)
  File "C:\Users\MLegrand\.local\pipx\venvs\pdm\lib\site-packages\pdm\models\candidates.py", line 416, in obtain
    raise CandidateNotFound(
pdm.exceptions.CandidateNotFound: No candidate is found for `connectorx` that matches the environment or hashes
```

## Expected behavior

```
✔ Install connectorx 0.3.1 successful
```

## Environment Information

```bash
# Paste the output of `pdm info && pdm info --env` below:

PDM version:
  4.8.0a2
Python Interpreter:
  D:\dev\myproject\.venv\Scripts\python.exe (3.11)
Project Root:
  D:/dev/myproject
Local Packages:

{
  "implementation_name": "cpython",
  "implementation_version": "3.11.2",
  "os_name": "nt",
  "platform_machine": "AMD64",
  "platform_release": "10",
  "platform_system": "Windows",
  "platform_version": "10.0.19045",
  "python_full_version": "3.11.2",
  "platform_python_implementation": "CPython",
  "python_version": "3.11",
  "sys_platform": "win32"
}
```

**Comment by m-legrand** [comment]
Actually I found the issue, and it has to do with the weird versioning of `connectorx`: 0.2.3, 0.3.2a1, 0.3.2a2, 0.3.2a3, 0.3.2a4, 0.3.2a5, 0.3.2a6, 0.3.2a7.

Somehow letting PDM decide on the version and then install it fails, so what I did was to pin it in my requirements by running `'connectorx==0.3.2a7'`. And it worked.

Any idea on how to deal with this case while setting a loose requirement letting this dependency update over time would be much appreciated. But having this workaround makes it much less urgent than I originally anticipated.
